nucypher/blockchain/eth/deployers.py

```diff
@@ -296,6 +311,63 @@ def deploy(self) -> dict:
         self.deployment_transactions = deployment_transactions
         return deployment_transactions
 
+    def upgrade(self, existing_secret_plaintext: bytes, new_secret_hash: bytes):
+
+        # Raise if not all-systems-go
+        self.check_deployment_readiness()
+        origin_args = {'from': self.deployer_address, 'gas': 5000000}  # TODO: Gas management
+
+        existing_bare_contract = self.blockchain.interface.get_contract_by_name(name=self.contract_name,
+                                                                                proxy_name=self.__proxy_deployer.contract_name,
+                                                                                use_proxy_address=False)
+        dispatcher_deployer = DispatcherDeployer(blockchain=self.blockchain,
+                                                 target_contract=existing_bare_contract,
+                                                 deployer_address=self.deployer_address,
+                                                 bare=True)  # acquire agency for the dispatcher itself.
+
+        # 2 - Deploy new version #
+        the_escrow_contract, deploy_txhash = self.blockchain.interface.deploy_contract(self.contract_name,
+                                                                                       self.token_agent.contract_address,
+                                                                                       *self.__economics.staking_deployment_parameters)
+
+        policy_agent = PolicyAgent(blockchain=self.blockchain)
+        _set_policy_manager_txhash = the_escrow_contract.functions.setPolicyManager(policy_agent.contract_address).transact(origin_args)
+        _set_policy_manager_receipt = self.blockchain.wait_for_receipt(_set_policy_manager_txhash)
```

Comment:
Main idea with proxy - use dispatcher because only dispatcher has right storage. Also minimum initialization process of contract (any contract) happens in constructor. So only deploy and retarget and no need to call `setPolicyManager` or `initialize` while upgrading

Reply:
:heavy_check_mark: